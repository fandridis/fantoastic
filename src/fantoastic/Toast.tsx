import React, { useEffect, useRef, useState, ReactNode } from "react";
import classnames from "classnames";

interface ToastProps {
  children: ReactNode;
  variant: string;
  duration: number;
  withCloseIcon: boolean;
  remove: (id: string) => void
}

interface TimerProps {
  pause: () => {},
  resume: () => {}
}

const Timer = function (this: any, callback: any, delay: any) {
  let timerId: any;
  let start: any;
  let remaining = delay;

  this.pause = function () {
    window.clearTimeout(timerId);
    remaining -= Date.now() - start;
  };

  this.resume = function () {
    start = Date.now();
    window.clearTimeout(timerId);
    timerId = window.setTimeout(callback, remaining);
  };

  this.resume();
};

function Toast(this: any, props: ToastProps) {
  const { children, variant, duration, withCloseIcon, remove } = props;
  const [show, setShow] = useState(true);
  const [isPaused, setIsPaused] = useState(false);
  const [timer, setTimer] = useState<TimerProps>();

  const removeRef = useRef(null);
  removeRef.current = remove;

  useEffect(() => {
    if (!duration) {
      return;
    }

    const timer = new (Timer as any)(() => {
      setShow(false);
    }, duration);

    setTimer(timer);

    return () => clearTimeout(timer);
  }, [duration]);

  const setToRemove = () => {
    setShow(false);
  };

  const onMouseOver = () => {
    if (isPaused || !timer) {
      return;
    }

    timer.pause();

    setIsPaused(true);
  };

  const onMouseLeave = () => {
    if (!isPaused || !timer) {
      return;
    }

    timer.resume();

    setIsPaused(false);
  };

  const onAnimationEnd = () => {
    if (!show) {
      removeRef.current();
    }
  };

  const wrapperClasses = classnames(
    `
    Fantoastic
    Fantoastic-variant--${variant}
  `,
    {
      "Fantoastic--visible": show,
      "Fantoastic--hidden": !show
    }
  );

  return (
    <div
      className={wrapperClasses}
      onAnimationEnd={onAnimationEnd}
      onMouseOver={onMouseOver}
      onMouseLeave={onMouseLeave}
    >
      <div className="Fantoastic-body">
        <div className="Fantoastic__content-wrapper">
          <div className="Fantoastic__content">{children}</div>
        </div>
        {withCloseIcon &&
          <div className="Fantoastic__closeIcon-wrapper" onClick={setToRemove}>
            <div className="Fantoastic__closeIcon Fantoastic__closeIcon--rounded"></div>
          </div>
        }
      </div>
    </div>
  );
}

export default Toast;
