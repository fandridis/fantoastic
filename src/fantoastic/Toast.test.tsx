import React, { ReactNode } from "react";
import { shallow } from "enzyme";
import Toast from './Toast';

function getPropsWithVariant(variant: string) {
  return {
    children: 'This is a fantastic toast!',
    variant: variant,
    duration: 1000,
    withCloseIcon: false,
    remove: (id: string) => { }
  }
}

function getPropsWithCloseIcon(withCloseIcon: boolean) {
  return {
    children: 'This is a fantastic toast!',
    variant: 'default',
    duration: 1000,
    ...(withCloseIcon && { withCloseIcon }),
    remove: (id: string) => { }
  }
}

describe("Variants: ", () => {
  describe("When render a default toast, ", () => {
    it("it should have the correct classes", () => {
      const props = getPropsWithVariant('default');
      const renderWrapper = () => shallow(<Toast {...props} />);

      expect(renderWrapper()).toMatchSnapshot();
    });
  });

  describe("When render a primary toast, ", () => {
    it("it should have the correct classes", () => {
      const props = getPropsWithVariant('primary');
      const renderWrapper = () => shallow(<Toast {...props} />);

      expect(renderWrapper()).toMatchSnapshot();
    });
  });

  describe("When render a success toast, ", () => {
    it("it should have the correct classes", () => {
      const props = getPropsWithVariant('success');
      const renderWrapper = () => shallow(<Toast {...props} />);

      expect(renderWrapper()).toMatchSnapshot();
    });
  });

  describe("When render a danger toast, ", () => {
    it("it should have the correct classes", () => {
      const props = getPropsWithVariant('danger');
      const renderWrapper = () => shallow(<Toast {...props} />);

      expect(renderWrapper()).toMatchSnapshot();
    });
  });
});


describe("CloseIcon: ", () => {
  describe("When withCloseIcon is not set, ", () => {
    it("should not render a closeIcon element", () => {
      const props = getPropsWithCloseIcon(false);
      const renderWrapper = () => shallow(<Toast {...props} />);

      expect(renderWrapper()).toMatchSnapshot();
    });
  });

  describe("When withCloseIcon is set, ", () => {
    it("should render a closeIcon element", () => {
      const props = getPropsWithCloseIcon(true);
      const renderWrapper = () => shallow(<Toast {...props} />);

      expect(renderWrapper()).toMatchSnapshot();
    });
  });
});
