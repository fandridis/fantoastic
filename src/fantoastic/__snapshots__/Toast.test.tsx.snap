// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseIcon:  When withCloseIcon is not set,  should not render a closeIcon element 1`] = `
<div
  className="
    Fantoastic
    Fantoastic-variant--default
   Fantoastic--visible"
  onAnimationEnd={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="Fantoastic-body"
  >
    <div
      className="Fantoastic__content-wrapper"
    >
      <div
        className="Fantoastic__content"
      >
        This is a fantastic toast!
      </div>
    </div>
  </div>
</div>
`;

exports[`CloseIcon:  When withCloseIcon is set,  should render a closeIcon element 1`] = `
<div
  className="
    Fantoastic
    Fantoastic-variant--default
   Fantoastic--visible"
  onAnimationEnd={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="Fantoastic-body"
  >
    <div
      className="Fantoastic__content-wrapper"
    >
      <div
        className="Fantoastic__content"
      >
        This is a fantastic toast!
      </div>
    </div>
    <div
      className="Fantoastic__closeIcon-wrapper"
      onClick={[Function]}
    >
      <div
        className="Fantoastic__closeIcon Fantoastic__closeIcon--rounded"
      />
    </div>
  </div>
</div>
`;

exports[`Variants:  When render a danger toast,  it should have the correct classes 1`] = `
<div
  className="
    Fantoastic
    Fantoastic-variant--danger
   Fantoastic--visible"
  onAnimationEnd={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="Fantoastic-body"
  >
    <div
      className="Fantoastic__content-wrapper"
    >
      <div
        className="Fantoastic__content"
      >
        This is a fantastic toast!
      </div>
    </div>
  </div>
</div>
`;

exports[`Variants:  When render a default toast,  it should have the correct classes 1`] = `
<div
  className="
    Fantoastic
    Fantoastic-variant--default
   Fantoastic--visible"
  onAnimationEnd={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="Fantoastic-body"
  >
    <div
      className="Fantoastic__content-wrapper"
    >
      <div
        className="Fantoastic__content"
      >
        This is a fantastic toast!
      </div>
    </div>
  </div>
</div>
`;

exports[`Variants:  When render a primary toast,  it should have the correct classes 1`] = `
<div
  className="
    Fantoastic
    Fantoastic-variant--primary
   Fantoastic--visible"
  onAnimationEnd={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="Fantoastic-body"
  >
    <div
      className="Fantoastic__content-wrapper"
    >
      <div
        className="Fantoastic__content"
      >
        This is a fantastic toast!
      </div>
    </div>
  </div>
</div>
`;

exports[`Variants:  When render a success toast,  it should have the correct classes 1`] = `
<div
  className="
    Fantoastic
    Fantoastic-variant--success
   Fantoastic--visible"
  onAnimationEnd={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="Fantoastic-body"
  >
    <div
      className="Fantoastic__content-wrapper"
    >
      <div
        className="Fantoastic__content"
      >
        This is a fantastic toast!
      </div>
    </div>
  </div>
</div>
`;
